R/check_package.R

```diff
@@ -0,0 +1,87 @@
+#' check_package
+#'
+#' This function perform checks on a package before publishing
+#' @param mn MNode
+#' @param resource_map resource_map pid
+check_package <- function(mn, resource_map) {
+
+    stopifnot(is(mn, "MNode"))
+    stopifnot(is.character(resource_map))
+
+    cat("\nRunning Check on Package...")
+    pkg <- arcticdatautils::get_package(mn, resource_map, file_names = TRUE)
+    eml <- EML::read_eml(rawToChar(getObject(mn, pkg$metadata)))
+
+    pids <- vapply(unlist(eml_get(eml, "url")), function(x) {
+        stringr::str_match_all(x, "https?:\\/\\/.+\\.dataone\\.org\\/cn\\/v\\d\\/resolve\\/(.+)")[[1]][2]
+    }, "", USE.NAMES = FALSE)
+
+    if(length(pids) != length(pkg$data)){
+        stop("Number of dataTables and otherEntities is ", length(pids),
+             ". Number of data in package is ", length(pids))
+    }
+
+    pkg_in_pid <- (pkg$data %in% pids)
+    if(!all(pkg_in_pid)){
+        stop("The following pids are in the package and not in the EML.", paste("\n",pkg$data[!pkg_in_pid]))
+    }
+
+    cat("\nThe pids in the package are the same as the pids in the EML.")
+
+    objectNames <- unlist(eml_get(eml, "objectName"))
+    pkg_Names <- names(pkg$data)
+
+    Names_in_EML <- (pkg_Names %in% objectNames)
+    if(!all(Names_in_EML)){
+        stop("The following file names are listed in the package and not the EML", paste("\n",pkg_Names[!Names_in_EML]))
+    }
+
+    cat("\nThe data names in the package are the same as the data names in the EML.")
+
+    creators <- eml@dataset@creator
+    if (length(creators) == 0) {
+        stop("The EML needs a creator.")
+    }
+
+    creator_ORCIDs <- unlist(eml_get(creators,"userId"))
+
+    isORCID <-  grepl("https:\\/\\/orcid.org\\/[[:digit:]]{4}-[[:digit:]]{4}-[[:digit:]]{4}-[[:digit:]]{4}",creator_ORCIDs)
+
+    if(!all(isORCID)){
+        stop(creator_ORCIDs[!isORCID], " is not of the form https://orcid.org/AAAA-BBBB-CCCC-DDDD")
+    }
+
+    if(length(creator_ORCIDs)!=length(creators)){
+        warning("Each Creator should have an ORCID.")
+    }
+
+    creator_ORCIDs <- sub("https://","http://",creator_ORCIDs,fixed = T)
+
+
+    all_pids <- c(pkg$metadata, pkg$resource_map, pkg$data)
+    permissions <- c("read","write","changePermission")
+
+    for (pid in all_pids) {
+        sysmeta <- dataone::getSystemMetadata(mn, pid)
```

Comment:
You could do this with a solr query in one call outside of the for loop.  I can send you one if you want.

Reply:
```
#' Calculate the total size of the Objects in a Data Package
#'
#' @param node (MNode/CNode) The Node to query for Object sizes
#' @param resource_map_pid (character) The identifier of the Data Package's Resource Map
#' @param formatType (character) Optional. Filter to just Objects of the given formatType. One of METADATA, RESOURCE, or DATA or * for all types
#'
#' @author Bryce Mecum 
#'
#' @return (numeric) The sum of all Object sizes in the Data Package
get_package_size <- function(node, package_identifier, formatType = "*") {
  size_query <- dataone::query(node,
                               paste0("q=resourceMap:\"",
                                      package_identifier,
                                      "\"+AND+formatType:",
                                      formatType, "&fl=size"),
                               as = "data.frame")

  if (nrow(size_query) == 0) {
    return(0)
  }

  sum(as.integer(size_query$size))
}
```

Reply:
I like this approach in theory, but don't know if it actually saves from doing loops.